File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.java
class Solution {
    public boolean search(int[] arr, int target) {
        int pivot = findPiv(arr);
        boolean ans = BinarySearch(arr, target, 0, pivot);
        if(ans == false){
            return BinarySearch(arr, target, pivot+1, arr.length - 1);
        }
        return ans;
    }
    int findPiv(int[] arr){
        int start = 0;
        int end = arr.length - 1;
        while(start <= end){
            int mid = start + (end - start) / 2;
            if(mid < end && arr[mid] > arr[mid + 1]) return mid ;
            if(start < mid && arr[mid] < arr[mid - 1]) return mid-1;
             
            if(arr[mid] == arr[start] && arr[mid] == arr[end]){
                if(start < mid && arr[start] > arr[start + 1]){
                    return start;
                }
                start++;
                if(mid < end && arr[end] < arr[end - 1]){
                    return end;
                }
                end--;
            }
            else if((arr[start] < arr[mid]) || (arr[start] == arr[mid] && arr[mid] > arr[end])){
                start = mid + 1;
            }
            else {
                end = mid - 1;
            }
            
        }
        return -1;

    }
    boolean BinarySearch(int[] arr, int target, int start, int end){
            while (end >= start){
                int mid = start + (end - start)/2;

                if(target < arr[mid]) end = mid - 1;

                else if(target > arr[mid]) start = mid + 1;

                else return true;

            }
            return false;
    }
}

File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.c
int findPiv(int *arr, int n){
    for(int i = 0; i < n - 1; i++) if(arr[i] > arr[i+1]) return i;
    return -1;
}
bool bs(int *arr, int start, int end, int target){
    while(start <= end){
        int mid = start + (end - start) / 2;
        if(arr[mid] == target) return true;
        else if(arr[mid] > target) end = mid - 1;
        else start = mid + 1;
    }
    return false;
}
bool search(int* arr, int n, int target) {
    int piv = findPiv(arr, n);
    if(n == 1 && arr[0] == target || arr[0] == target) return true;
    else if(piv == 0) return bs(arr, piv + 1, n - 1, target); 
    else if(bs(arr, 0, piv, target) == true) return true;
    else if(bs(arr, piv+1, n-1, target) == true) return true;
    return false; 
}